src/itwinai/torch/distributed.py

```diff
@@ -1010,3 +1012,177 @@ def gather(self, tensor: torch.Tensor, dst_rank: int = 0):
+                print("Ray initialized!")
+
+        except Exception as e:
+            raise RuntimeError(f"Error initializing Ray: {str(e)}")
+
+
+class RayDDPStrategy(RayDistributedStrategy):
+    def __init__(self) -> None:
+        super()._initialize_ray()
+
+    def init(self) -> None:
+        self.is_initialized = True
+
+    @check_initialized
+    def global_world_size(self) -> int:
+        return ray.train.get_context().get_world_size()
+
+    @check_initialized
+    def local_world_size(self) -> int:
+        return ray.train.get_context().get_local_world_size()
+
+    @check_initialized
+    def global_rank(self) -> int:
+        return ray.train.get_context().get_world_rank()
+
+    @check_initialized
+    def local_rank(self) -> int:
+        return ray.train.get_context().get_local_rank()
+
+    @check_initialized
+    def distributed(
+        self,
+        model: nn.Module,
+        optimizer: Optimizer,
+        lr_scheduler: Optional[LRScheduler] = None,
+    ) -> Tuple[Module | Optimizer | LRScheduler | None]:
+        model = ray.train.torch.prepare_model(model)
+
+        return model, optimizer, lr_scheduler
+
+    @check_initialized
+    def create_dataloader(
+        self,
+        dataset: Dataset[T_co],
+        batch_size: Optional[int] = 1,
+        shuffle: Optional[bool] = None,
+        sampler: Union[Sampler, Iterable, None] = None,
+        batch_sampler: Union[Sampler[List], Iterable[List], None] = None,
+        num_workers: int = 0,
+        collate_fn: Optional[_collate_fn_t] = None,
+        pin_memory: bool = False,
+        drop_last: bool = False,
+        timeout: float = 0,
+        worker_init_fn: Optional[_worker_init_fn_t] = None,
+        multiprocessing_context=None,
+        generator=None,
+        *,
+        prefetch_factor: Optional[int] = None,
+        persistent_workers: bool = False,
+        pin_memory_device: str = "",
+    ):
+        dataloader = super().create_dataloader(
+            dataset=dataset,
+            batch_size=batch_size,
+            sampler=sampler,
+            shuffle=shuffle,
+            num_workers=num_workers,
+            collate_fn=collate_fn,
+            pin_memory=pin_memory,
+            drop_last=drop_last,
+            timeout=timeout,
+            worker_init_fn=worker_init_fn,
+            multiprocessing_context=multiprocessing_context,
+            generator=generator,
+            prefetch_factor=prefetch_factor,
+            persistent_workers=persistent_workers,
+            pin_memory_device=pin_memory_device,
+        )
+
+        return ray.train.torch.prepare_data_loader(dataloader)
+
+    def clean_up(self) -> None:
+        pass
+
+    def allgather_obj(self, obj: Any) -> List[Any]:
+        pass
+
+    def gather_obj(self, obj: Any, dst_rank: int = 0) -> List[Any]:
+        pass
+
+    def gather(self, tensor: torch.Tensor, dst_rank: int = 0) -> List | None:
+        pass
+
+
+class RayDeepSpeedStrategy(RayDistributedStrategy):
```

Comment:
Would it be possible for this strategy to inherit "DeepSpeedStrategy" (or whatever it is called) instead? Would that make any sense at all?

Reply:
Yeah I thought about this, it would definitely save some lines of code as many functions are exactly the same, e.g. `distributed` and `global/local_world_size`, etc. On the other hand we would have to then move the ray initialisation logic into the RayDeepSpeedStrategy.. Also, at least imo, it's nicer to have all ray strategies uniformly inheriting from the same strategy, but that is more a design choice, both would work. I would however, if we changed this, remove the RayDistributedStrategy altogether and have all ray strategies inherit from their pure torch counterpart. There is also some frameworks that ray supports and itwinai currently does not (huggingface transformers/ accelerate for example), that I thought we might want to implement in the future, in which case they would not have an itwinai torch strategy counterpart

Reply:
I would relax a bit the structure by transforming `_initialize_ray` method in a function and removing the `RayDistributedStrategy`.

I don't think that there is much in common between DDP and RayDDP, so RayDDP could inherit directly from TorchDistributedStrategy. 

Regarding deepspeed, I would try to reuse as much as possible from the DeepSpeedStrategy class already there, thus having RayDeepSpeedStrategy to inherit from DeepSpeedStrategy.

I agree that it would be nice to have an overarching Ray class, and we could actually have it (RayDeepSpeedStrategy  could inherit both from RayDistributedStrategy and DeepSpeedStrategy). But maybe it's complicating things a bit?

What do you think?

Reply:
I definitely see your point - we could move the inheritance, but I would keep it uniform and have the ray strategies inherit from their torch counterpart. The RayDDPStrategy would be overwriting most of the functions, but (to me, at least) it would be better style to have it uniform, otherwise I think the inheritance would be confusing, especially when we add more strategies. 

And then for the second point, I think it's hard to tell now if having a separate RayDistributedStrategy makes sense, because 
a. there is at least two more strategies that I think would be cool to add (horovod and huggingface accelerate), and I don't know yet which methods they need and if anything else could be consolidated in a shared parent class
b. I have not had a chance to look at some functions, such as `gather` and `clean_up`, so same for these

So I would suggest to leave the RayDistributedStrategy for now, and if you think it would be better, I can add multi-inheritance to make the RayDDPStrategy inherit from the DDPStrategy and the RayDeepSpeedStrategy to inherit from the DeepSpeedStrategy.

Reply:
We can still refine it in the future, if we add more strategies 😌